utils/runtime-fuzzer/src/gear_calls.rs

```diff
@@ -0,0 +1,409 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! [`GearCalls`] implementation.
+//!
+//! [`GearCalls`]'s purpose is to lazy evaluate [`GearCall`]s using provided [`Config`]
+//! (see the [`GearCalls`] docs for the more detailed information about usage).
+//!
+//! [`Config`] is basically a set of [`ExtrinsicGenerator`]s, each of them is capable of
+//! generating one particular extrinsic's parameters, for example - [`UploadProgramGenerator`]
+//! or [`SendMessageGenerator`].
+//!
+//! Some of the extrinsics require access to the mailbox messages, such as `SendReply`
+//! and `ClaimValue` ones. This access is provided by the [`MailboxProvider`] trait
+//! which must be implemented by the callling side and passed into generators which require it.
+
+use crate::{GearCall, SendMessageArgs, UploadProgramArgs};
+use arbitrary::{Error, Result, Unstructured};
+use gear_call_gen::{ClaimValueArgs, SendReplyArgs};
+use gear_core::ids::{CodeId, MessageId, ProgramId};
+use gear_utils::NonEmpty;
+use gear_wasm_gen::{
+    EntryPointsSet, ParamType, StandardGearWasmConfigsBundle, SysCallName,
+    SysCallsInjectionAmounts, SysCallsParamsConfig,
+};
+
+/// Maximum payload size for the fuzzer - 512 KiB.
+const MAX_PAYLOAD_SIZE: usize = 512 * 1024;
+static_assertions::const_assert!(MAX_PAYLOAD_SIZE <= gear_core::message::MAX_PAYLOAD_SIZE);
+
+/// This trait provides ability for [`ExtrinsicGenerator`]s to fetch messages
+/// from mailbox, for example [`UploadProgramGenerator`] and
+/// [`ClaimValueGenerator`] use it.
+pub(crate) trait MailboxProvider {
+    fn fetch_messages(&self) -> Vec<MessageId>;
```

Comment:
ProgramId of dest of the messages?

Reply:
This trait represents mailbox of a concrete user, so we can configure user address when creating generators